Fix label smoothing incompatibility with multi-label classification

avchauzov opened the pull request:
# What does this PR do?

Fixes a RuntimeError when using `label_smoothing_factor > 0` with multi-label classification.

Previously, users would get a confusing error:
`RuntimeError: gather(): Expected dtype int32/int64 for index`

This PR adds a clear validation check in `LabelSmoother` that detects this incompatible combination and raises a helpful error message instead:
```
ValueError: Label smoothing is not compatible with multi-label classification.
Multi-label classification uses one-hot encoded labels (float tensors),
but label smoothing requires integer class indices.
Please set label_smoothing_factor=0.0 in your TrainingArguments.
```

**Changes:**
- Add validation in `LabelSmoother.__call__()` to detect multi-label classification
- Raise clear `ValueError` with actionable solution  
- Add test case to prevent regression

**Testing:**
- ✅ All existing tests pass
- ✅ New test covers the fix
- ✅ Original reproduction case now shows clear error

Fixes #40258

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?
@zach-huggingface @SunMarc @qgallouedec - trainer related changes

avchauzov committed:
Fix label smoothing incompatibility with multi-label classification (#40258)

SunMarc commented on src/transformers/trainer_pt_utils.py:
```diff
@@ -555,6 +555,16 @@ def __call__(self, model_output, labels, shift_labels=False):
             logits = logits[..., :-1, :].contiguous()
             labels = labels[..., 1:].contiguous()
 
+        # Check for multi-label classification incompatibility
+        if labels.dtype == torch.float and labels.dim() > 1:
```
i feel like this seems not super robust check for multi-labels, can you try to improve this ?

SunMarc replied:
option 2 seems much better to me ! Instead of raising an error value, we can just warn that we've set self.label_smoother = None as we can't use it with multi-label

avchauzov replied:
@SunMarc Thanks for the feedback!

I've implemented option 2 as you suggested:

- Moved check to `Trainer.__init__()`
- `model.config.problem_type` - explicit configuration instead of tensor inference
- Warning instead of error raising
- Updated test

The original issue now shows a clear `UserWarning`:
```
Label smoothing is not compatible with multi-label classification.
Disabling label smoothing for this training run.
```

SunMarc reviewed (commented):
Thanks ! just a nit

avchauzov committed:
Improve label smoothing multi-label check based on reviewer feedback

- Move check from LabelSmoother to Trainer.__init__() for better architecture
- Use model.config.problem_type instead of tensor inference for robustness
- Warn and disable smoothing instead of raising error for better UX
- Update test to verify warning behavior

andreabosisio commented:
thank you @avchauzov @SunMarc for your work on this! I was wondering if it would actually be possible to implement label smoothing for multi-label classification tasks (#40301)

SunMarc reviewed (approved):
Thanks !

SunMarc commented:
> thank you @avchauzov @SunMarc for your work on this! I was wondering if it would actually be possible to implement label smoothing for multi-label classification tasks (https://github.com/huggingface/transformers/issues/40301)

I won't have the time but hopefully the community picks it !
